pkg/agent/monitortool/latency_store.go

```diff
@@ -0,0 +1,195 @@
+	store := &LatencyStore{
+		connectionMap: make(map[string]*Connection),
+		nodeInfoMap:   make(map[string]*corev1.Node),
+		nodeInformer:  nodeInformer,
+	}
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    store.OnNodeAdd,
+		UpdateFunc: store.OnNodeUpdate,
+		DeleteFunc: store.OnNodeDelete,
+	})
+
+	return store
+}
+
+func (l *LatencyStore) Run(stopCh <-chan struct{}) {
+	l.nodeInformer.Informer().Run(stopCh)
+}
+
+func (l *LatencyStore) OnNodeAdd(obj interface{}) {
+	node := obj.(*corev1.Node)
+	l.AddNodeToMap(node)
+}
+
+func (l *LatencyStore) OnNodeUpdate(oldObj, newObj interface{}) {
+	oldNode := oldObj.(*corev1.Node)
+	node := newObj.(*corev1.Node)
+	l.UpdateNodeByKey(oldNode.Name, node)
+}
+
+func (l *LatencyStore) OnNodeDelete(obj interface{}) {
+	node := obj.(*corev1.Node)
+	l.DeleteNodeByKey(node.Name)
+}
+
+func (l *LatencyStore) AddConnToMap(connKey string, conn *Connection) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	// Judge if the connection is already in the map
+	_, found := l.connectionMap[connKey]
+	if !found {
+		conn.CreatedAt = conn.LastUpdated
+	}
+
+	l.connectionMap[connKey] = conn
+}
+
+func (l *LatencyStore) GetConnByKey(connKey string) (*Connection, bool) {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	conn, found := l.connectionMap[connKey]
+
+	return conn, found
+}
+
+func (l *LatencyStore) DeleteConnByKey(connKey string) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	delete(l.connectionMap, connKey)
+}
+
+func (l *LatencyStore) UpdateConnByKey(connKey string, conn *Connection) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	// Judge if the connection is already in the map
+	_, found := l.connectionMap[connKey]
+	if !found {
+		conn.CreatedAt = conn.LastUpdated
+	}
+
+	l.connectionMap[connKey] = conn
+}
+
+func (l *LatencyStore) ListConns() map[string]*Connection {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	return l.connectionMap
+}
+
+func (l *LatencyStore) AddNodeToMap(node *corev1.Node) {
+	l.nodeInfoMap[node.Name] = node
+}
+
+func (l *LatencyStore) GetNodeByKey(nodeKey string) (*corev1.Node, bool) {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	node, found := l.nodeInfoMap[nodeKey]
+
+	return node, found
+}
+
+func (l *LatencyStore) DeleteNodeByKey(nodeKey string) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	// Delete the node from the node info map
+	delete(l.nodeInfoMap, nodeKey)
+	// Delete the node from the connection map
+	delete(l.connectionMap, nodeKey)
+}
+
+func (l *LatencyStore) UpdateNodeByKey(nodeKey string, node *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	l.nodeInfoMap[nodeKey] = node
+}
+
+func (l *LatencyStore) ListNodes() []corev1.Node {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	nodes := make([]corev1.Node, 0, len(l.nodeInfoMap))
+	for _, node := range l.nodeInfoMap {
+		nodes = append(nodes, *node)
+	}
+
+	return nodes
+}
+
+func (l *LatencyStore) ListNodeIPs() map[string]string {
```

Comment:
for the Antrea encap mode, there should definitely be a way to use the antrea-gw0 IP as the destination IP instead

Reply:
IP configs seem to be stored here, I will update it in next commit.`pkg/agent/controller/noderoute/node_route_controller.go`